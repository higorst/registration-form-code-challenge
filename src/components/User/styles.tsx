import styled from 'styled-components/native'
import { Dimensions, ViewProps } from 'react-native'
import { Colors } from '../../styles/Colors'

interface PropsTypeUser extends ViewProps {
    color: string
}
interface CardType extends ViewProps {
    borderColor: string
}
interface TouchableType extends ViewProps {
    color: string
}
interface MessageType extends ViewProps {
    italic?: boolean
    marginLeft?: boolean
}

export const Message = styled.Text`
    font-size: 16px;
    text-align: center;
    max-width: ${`${Dimensions.get('window').width * 0.6}px`};

    font-style: ${(props: MessageType) => props.italic ? 'italic' : 'normal' };
    margin-left: ${(props: MessageType) => props.marginLeft ? 
                    `${Dimensions.get('window').width * 0.09}px` : '0px' };
    color: ${Colors.white};
`
export const Card = styled.View`
    flex: 1;
    flex-direction: column;
    justify-content: space-between;

    padding: 10px;
    margin: 5px;

    border-width: 2px;
    border-color: ${(props: CardType) => props.borderColor };
    border-radius: 10px;
`

export const Row = styled.View<ViewProps>`
    flex: 1;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    margin: 3px;
`
export const TypeUser = styled.View`
    width: 20px;
    height: 20px;
    border-radius: 20px;

    background-color: ${(props: PropsTypeUser) => props.color };
`
export const NameUser = styled.Text`
    color: ${Colors.white};
    width: ${`${Dimensions.get('window').width * 0.78}px`};

    text-align: left;  
    font-size: 18px;
    font-weight: bold;
    margin-left: 15px;
`