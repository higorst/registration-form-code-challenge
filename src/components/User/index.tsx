import React, { useState } from 'react'
import { Colors } from '../../styles/Colors';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

import { Card, UserAndType, DocumentAndButton, TypeUser, NameUser, DocumentUser, Touchable } from './styles'
import Popup from '../Popup';
import RegexControllers from '../../controllers/Regexcontrollers';

// Redux
import { connect } from "react-redux";
import { SAGA_DELETE } from '../../redux/sagas/types';

import { Constants } from '../../constants';

interface UserType {
    user: {
        name: string
        document: number
        type: string
    }

    delete?: any
}
const User = (props: UserType) => {

    const [showPopup, setShowPopup] = useState(false)

    return (
        <Card
            borderColor={(props.user.type === Constants.typeUserDataIndividual) ? Colors.individual : Colors.business}
        >
            <Popup
                visible={showPopup}
                onCancel={() => setShowPopup(false)}
                user_name={props.user.name}
                onOk={() => {
                    setShowPopup(false)
                    props.delete({
                        name: props.user.name,
                        document: props.user.document,
                        type: props.user.type
                    })
                }}
            />
            <UserAndType>
                <TypeUser
                    color={(props.user.type === Constants.typeUserDataIndividual) ? Colors.individual : Colors.business}
                />
                <NameUser>{props.user.name}</NameUser>
            </UserAndType>
            <DocumentAndButton>
                <DocumentUser>{
                    RegexControllers.handleRegex(props.user.type === Constants.typeUserDataIndividual, props.user.document.toString(), true)
                }</DocumentUser>
                <Touchable
                    onPress={() => setShowPopup(true)}
                >
                    <Icon
                        name={Constants.iconButtonRemoveUser}
                        size={20}
                        color={Colors.white}
                    />
                </Touchable>
            </DocumentAndButton>
        </Card>
    )
}

const mapDispatchToProps = (dispatch: any) => ({
    delete: (data: UserType) => dispatch({ type: SAGA_DELETE, register: data }, dispatch),
})
export default connect(
    mapDispatchToProps,
)(User)