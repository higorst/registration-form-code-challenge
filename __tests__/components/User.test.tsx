import React from 'react';
import renderer from 'react-test-renderer';

// REDUX
import { Provider } from 'react-redux';
import store from '../../src/redux'

import User from '../../src/components/User/index'

jest.useFakeTimers()
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper')

describe('<User />', () => {
    it('render User Component correctly', () => {
        const tree = renderer.create(
            <Provider store={store}>
                <User 
                    user={
                        {
                            name: 'Snapshot',
                            document: 11111111111,
                            type: 'individual',
                        }
                    }
                />
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});