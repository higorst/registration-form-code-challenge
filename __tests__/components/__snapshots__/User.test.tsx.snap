// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<User /> render User Component correctly 1`] = `
<View
  borderColor="#ff6a00"
  style={
    Array [
      Object {
        "borderColor": "#ff6a00",
        "borderRadius": 10,
        "borderWidth": 2,
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "space-between",
        "marginBottom": 5,
        "marginLeft": 5,
        "marginRight": 5,
        "marginTop": 5,
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <View
      color="#ff6a00"
      style={
        Array [
          Object {
            "backgroundColor": "#ff6a00",
            "borderRadius": 20,
            "height": 20,
            "width": 20,
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 16,
            "fontWeight": "bold",
            "marginLeft": 15,
            "textAlign": "center",
          },
        ]
      }
    >
      Snapshot
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "space-between",
          "marginBottom": 2,
          "marginLeft": 35,
          "marginRight": 2,
          "marginTop": 2,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 16,
            "fontStyle": "italic",
            "textAlign": "center",
          },
        ]
      }
    >
      CPF: 111.111.111-11
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#da4f42",
          "borderRadius": 5,
          "justifyContent": "center",
          "opacity": 1,
          "paddingBottom": 5,
          "paddingLeft": 5,
          "paddingRight": 5,
          "paddingTop": 5,
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;
